### PR TITLE
Add input_grayscale_image_pyramid, issue #354

### PR DESCRIPTION
**Pull request by facug91**
Issue #354  
I'm doing some test with it, I'll be uploading a performance comparison tomorrow using it in the mmod (the training takes several hours).  
In the meantime I wanted to create the pull request, to be sure it's a good contribution, and to start reviewing the code.

### CODE REVIEW

**Comment by davisking** [comment]
Thanks for the PR.  The code looks correct.  You should use spaces instead of tabs though.  And if you could figure out how to reduce the code duplication that would be good too, like making a base class that contains most of the code and is inherited from.

**Comment by facug91** [comment]
Sorry about the tabs, I've fixed it already.  
About the hierarchy, what I could do is create a base class input_image_pyramid, which implement all methods except `to_tensor`, and all those related to serialization, which would go to the derived classes. Does this seem better for you?  
The problem is that `to_tensor` is a template member function, so I can not specialize it in derived classes.  
One more thing. If the above is good for you, what would be the correct way to document it in the abstract?

**Comment by davisking** [comment]
Sorry for the late reply.  Yeah, that sounds fine. Don't use any virtual functions, you can just publicly inherit from the base class to get all the methods.  As for `to_tensor`, yeah, that's not really sharable.  But their are parts of it that could be trivially moved into protected helper methods in the base class and called from derived classes.

I wouldn't mention this inheritance in the documentation at all.  It's an implementation detail users don't need to know about.

**Comment by facug91** [comment]
With this commit I've create a base clase `input_image_pyramid`, which has the common methods of the derived classes.

**Comment by facug91** [comment]
And now I've refactored to_tensor method, to move repeated code to base class.

**Comment by facug91** [comment]
Some benchmarking I've made:  

| | rgb | grayscale |
|-|------|--------------|
| 640x360 | 18.14 us | 16.05 us |
| 400x300 | 10.31 us | 9.22 us |

It is around 11% faster.

File: dlib/dnn/input.h [hunk 1]
```diff
@@ -591,30 +591,13 @@ namespace dlib
 // ----------------------------------------------------------------------------------------
 
     template <typename PYRAMID_TYPE>
-    class input_rgb_image_pyramid
+    class input_image_pyramid
```
Comment (davisking) [inline dlib/dnn/input.h]: This should probably go in a `detail` namespace to discourage users from touching it.

File: dlib/dnn/input.h [hunk 2]
```diff
@@ -623,8 +606,8 @@ namespace dlib
         void set_pyramid_outer_padding (unsigned long value) { pyramid_outer_padding = value; }
 
         bool image_contained_point (
-            const tensor& data,
-            const point& p
+                const tensor& data,
```
Comment (davisking) [inline dlib/dnn/input.h]: Fix tabbing here and other places.  Should be 4 spaces.

File: dlib/dnn/input.h [hunk 3]
```diff
@@ -633,45 +616,49 @@ namespace dlib
...
             for (auto i = ibegin; i != iend; ++i)
             {
                 DLIB_CASSERT(i->nr()==nr && i->nc()==nc,
-                    "\t input_rgb_image_pyramid::to_tensor()"
-                    << "\n\t All matrices given to to_tensor() must have the same dimensions."
-                    << "\n\t nr: " << nr
-                    << "\n\t nc: " << nc
-                    << "\n\t i->nr(): " << i->nr()
-                    << "\n\t i->nc(): " << i->nc()
+                             "\t input_grayscale_image_pyramid::to_tensor()"
```
Comment (davisking) [inline dlib/dnn/input.h]: `input_image_pyramid`

File: dlib/dnn/input.h [hunk 4]
```diff
@@ -690,11 +677,158 @@ namespace dlib
...
+        friend void to_xml(const input_grayscale_image_pyramid& item, std::ostream& out)
+        {
+            out << "<input_grayscale_image_pyramid"
+                <<"' pyramid_padding='"<<item.pyramid_padding
+                <<"' pyramid_outer_padding='"<<item.pyramid_outer_padding
+                <<"'/>";
+        }
+
+    private:
+        using input_image_pyramid<PYRAMID_TYPE>::pyramid_padding;
```
Comment (davisking) [inline dlib/dnn/input.h]: Is this needed?  Do you get an error in the friend functions without this?

File: dlib/dnn/input_abstract.h [hunk 5]
```diff
@@ -271,6 +271,155 @@ namespace dlib
...
+        >
+    class input_grayscale_image_pyramid
+    {
+        /*!
+            REQUIREMENTS ON PYRAMID_TYPE
+                PYRAMID_TYPE must be an instance of the dlib::pyramid_down template.
+
+            WHAT THIS OBJECT REPRESENTS
+                This input layer works with gray scale images of type matrix<unsigned char>.
+                It is identical to input except that it outputs a tensor containing a tiled
```
Comment (davisking) [inline dlib/dnn/input_abstract.h]: Say "identical to the input layer except..."

File: dlib/dnn/input.h (hunk 4 above)
Comment (facug91) [reply dlib/dnn/input.h]: I think it used to throw me a compilation error, but maybe the problem was somewhere else, and I thought it was that. I removed it and it's still working.

File: dlib/dnn/input_abstract.h (hunk 5 above)
Comment (facug91) [reply dlib/dnn/input_abstract.h]: Sorry, but I don't understand what you mean.

Comment (davisking) [reply dlib/dnn/input_abstract.h]: I just mean to replace "It is identical to input except that it outputs a tensor" with "It is identical to the input layer except that it outputs a tensor"

**Comment by davisking** [comment]
Sweet, looks good.  Thanks for the PR :)

**Pull request merged**